From f7ab7e2e536d9d401d4d6eae84c2e43573551451 Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Mon, 26 Aug 2019 09:41:58 -0700
Subject: [PATCH 0664/1851] v4l: xilinx: dma: Use early callback mode for low
 latency capture

commit 518f9ff4add02daacacc204ec73d6859346cd61a from
https://github.com/Xilinx/linux-xlnx.git

Rename early low latency capture early callback to
EARLY_CALLBACK_START_DESC .

Set EARLY_CALLBACK_START_DESC mode for the dmaengine so that
dmaengine driver can give early callback for buffer completion
at start of descriptor processing when low latency capture mode
is enabled.

Revert the old change to mark buffer as complete as we rely on
EARLY_CALLBACK_START_DESC to mark the buffer as done early
just before dmaengine starts processing the descriptor.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/dma/xilinx/xilinx_frmbuf.c         |  2 +-
 drivers/media/platform/xilinx/xilinx-dma.c | 13 +++++++------
 include/linux/dma/xilinx_frmbuf.h          |  7 ++++---
 3 files changed, 12 insertions(+), 10 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_frmbuf.c b/drivers/dma/xilinx/xilinx_frmbuf.c
index a135af17129b..f784eda90156 100644
--- a/drivers/dma/xilinx/xilinx_frmbuf.c
+++ b/drivers/dma/xilinx/xilinx_frmbuf.c
@@ -1096,7 +1096,7 @@ static void xilinx_frmbuf_start_transfer(struct xilinx_frmbuf_chan *chan)
 				struct xilinx_frmbuf_tx_descriptor,
 				node);
 
-	if (desc->earlycb == EARLY_CALLBACK_LOW_LATENCY) {
+	if (desc->earlycb == EARLY_CALLBACK_START_DESC) {
 		dma_async_tx_callback callback;
 		void *callback_param;
 
diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index c16dbcc53072..2c1fcea932b8 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -631,12 +631,13 @@ static void xvip_dma_buffer_queue(struct vb2_buffer *vb)
 	list_add_tail(&buf->queue, &dma->queued_bufs);
 	spin_unlock_irq(&dma->queued_lock);
 
-	/* Low latency capture: Early release of buffers to application */
-	if (dma->low_latency_cap) {
-		desc->callback = NULL;
-		xvip_dma_complete(buf);
-	}
-
+	/*
+	 * Low latency capture: Give descriptor callback at start of
+	 * processing the descriptor
+	 */
+	if (dma->low_latency_cap)
+		xilinx_xdma_set_earlycb(dma->dma, desc,
+					EARLY_CALLBACK_START_DESC);
 	dmaengine_submit(desc);
 
 	if (vb2_is_streaming(&dma->queue))
diff --git a/include/linux/dma/xilinx_frmbuf.h b/include/linux/dma/xilinx_frmbuf.h
index 5307113efd5b..54aa21e6fbec 100644
--- a/include/linux/dma/xilinx_frmbuf.h
+++ b/include/linux/dma/xilinx_frmbuf.h
@@ -15,9 +15,10 @@
 #include <linux/dmaengine.h>
 
 /* Modes to enable early callback */
-#define EARLY_CALLBACK			BIT(1) /* To avoid first frame delay */
-#define EARLY_CALLBACK_LOW_LATENCY	BIT(2) /* Low latency mode */
-
+/* To avoid first frame delay */
+#define EARLY_CALLBACK			BIT(1)
+/* Give callback at start of descriptor processing */
+#define EARLY_CALLBACK_START_DESC	BIT(2)
 /**
  * enum vid_frmwork_type - Linux video framework type
  * @XDMA_DRM: fourcc is of type DRM
-- 
2.31.1

